config/auth_metadata_service.go

```diff
@@ -0,0 +1,169 @@
+// is not set or set to a version that the server doesn't support.
+//
+// The server MUST guarantee stable sessions per URL path. That is, if the
+// server returns a token for a Host on a given URL path, it MUST continue to return
+// tokens for the same Host.
+//
+// The server MUST return a 4xx response if the Host passed in the "X-Databricks-Host"
+// header doesn't match the token.
+//
+// The server is expected to return a JSON response with the following fields:
+//
+// - access_token: The requested access token.
+// - token_type: The type of token, which is a "Bearer" access token.
+// - expires_on: Unix timestamp when the access token expires.
+
+type MetadataServiceCredentials struct{}
+
+func (c MetadataServiceCredentials) Name() string {
+	return "metadata-service"
+}
+
+func (c MetadataServiceCredentials) Configure(ctx context.Context, cfg *Config) (func(*http.Request) error, error) {
+	if cfg.MetadataServiceURL == "" || cfg.Host == "" {
+		return nil, nil
+	}
+
+	parsedMetadataServiceURL, err := url.Parse(cfg.MetadataServiceURL)
+	if err != nil {
+		return nil, fmt.Errorf("invalid auth server URL: %w", err)
+	}
+
+	// only allow localhost URLs
+	if parsedMetadataServiceURL.Hostname() != "localhost" && parsedMetadataServiceURL.Hostname() != "127.0.0.1" {
+		return nil, fmt.Errorf("invalid auth server URL: %s", cfg.MetadataServiceURL)
+	}
+
+	ms := metadataService{
+		metadataServiceURL: parsedMetadataServiceURL,
+		config:             cfg,
+	}
+
+	response, err := ms.Get()
+	if err != nil {
+		return nil, err
+	}
+
+	if response == nil {
+		return nil, nil
+	}
+
+	ts := metadataServiceTokenSource{
+		metadataService: &ms,
+		databricksHost:  cfg.Host,
+	}
+	return refreshableVisitor(&ts), nil
+}
+
+type metadataService struct {
+	metadataServiceURL *url.URL
+	config             *Config
+}
+
+// performs a request to the metadata service and returns the token
+func (s metadataService) Get() (*serverResponse, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), metadataServiceTimeout)
+	defer cancel()
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, s.metadataServiceURL.String(), nil)
+	if err != nil {
+		return nil, fmt.Errorf("token request: %w", err)
+	}
+
+	req.Header.Add(MetadataServiceVersionHeader, MetadataServiceVersion)
+	req.Header.Add(MetadataServiceHostHeader, s.config.Host)
+
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return nil, fmt.Errorf("token response: %w", err)
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode == http.StatusNotFound {
+		return nil, nil
+	}
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, fmt.Errorf("token read: %w", err)
+	}
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("token error: %s", body)
+	}
+
+	var response serverResponse
+	if err := json.Unmarshal(body, &response); err != nil {
+		return nil, fmt.Errorf("token parse: %w", err)
+	}
```

Comment:
this looks like a duplicate from 
https://github.com/databricks/databricks-sdk-go/blob/main/config/auth_azure_msi.go#L125-L145

for example, can we consider creating the token source, calling `.Token()` on it to verify if it's correctly initialized and then return a refreshable visitor, like Azure CLI provider does?

https://github.com/databricks/databricks-sdk-go/blob/main/config/auth_azure_cli.go#L34-L35